.github/workflows/build.yml

```
@@ -75,14 +82,25 @@ jobs:
       matrix:
         os: [ ubuntu-20.04, macos-12, windows-2022 ]
         python-version: [ 3.9, "3.10", "3.11", "3.12" ]
+        base: [ true, false ]
+        exclude:
+          - os: ubuntu-20.04
+            python-version: 3.12
+            base: true
+        include:
+          - os: ubuntu-20.04
+            python-version: 3.12
+            base: true
+            single_action_config: "True"
 
     env:
       RUNNER_OS: ${{ matrix.os }}
       PYTHON_VERSION: ${{ matrix.python-version }}
       SINGLE_ACTION_CONFIG: "${{ matrix.single_action_config == 'True' }}"
+      PIP_INSTALL: "${{ matrix.python-version == '3.12' }}"
       PYPI_SERVER: ${{ secrets.PYPI_SERVER }}
       PYPI_USER: ${{ secrets.PYPI_USER }}
-      PYPI_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
+      PYPI_PASSWORD: ${{ matrix.base && secrets.PYPI_BASE_PASSWORD || secrets.PYPI_PASSWORD }}
```

if my understanding is correct, we want to check if matrix.base is True and if so, we set PYPI_PASSWORD to be ``secrets.PYPI_BASE_PASSWORD``, is this code implementing this functionality?

---

I don't actually know what y'all have set up with the secrets, but _I assume_ based on the way the job was already coded that there are separate passwords for the `cvxpy` and `cvxpy-base` packages being submitted to pypi.  This is supposed to condition which one of those ends up in the `PYPI_PASSWORD` environment variable.

Several other places in this file that check for true/false values seem to assume that the matrix outputs are strings.  I was hoping maybe they can be just directly used as booleans, but I'd want to test that!

---

I understand what you mean here, but the current code has a condition on ``matrix.base && secrets.PYPI_BASE`` which doesn't seem to set the ``PYPI_PASSWORD`` to the correct value (doesn't it return a boolean instead now?)

---

I think (as is true with short-circuit logical operators in python) the JS evaluation of these template fields doesn't coerce to boolean values.

See, for example, the existing code here:
https://github.com/cvxpy/cvxpy/blob/273e0cfa65a90440b0e5cf05164a76a501edf6c0/.github/workflows/build.yml#L41

Or compare to python:
```python
>>> False and "hi" or "bye"
'bye'
>>> True and "hi" or "bye"
'hi'
```

---

ohhh I see, it's some javascript trick.. looks good then

---

One thing I realized from this conversation that I do still want to verify is whether matrix values can actually be native booleans, or if I might need this to be
```suggestion
      PYPI_PASSWORD: ${{ matrix.base == 'true' && secrets.PYPI_BASE_PASSWORD || secrets.PYPI_PASSWORD }}
```
I'm going to push a commit that tests this on the regular "build" section of the workflow so that it's something we can test before merging the PR.
(Actually, we should talk more about what we _can_ do to test this more thoroughly before merging?)